pyro/distributions/hmm.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.distributions import constraints
+
+from pyro.distributions.torch_distribution import TorchDistribution
+from pyro.distributions.util import broadcast_shape
+
+
+def _logmatmulexp(x, y):
+    """
+    Numerically stable version of ``(x.log() @ y.log()).exp()``.
+    """
+    x_shift = x.max(-1, keepdim=True)[0]
+    y_shift = y.max(-2, keepdim=True)[0]
+    xy = torch.matmul((x - x_shift).exp(), (y - y_shift).exp()).log()
+    return xy + x_shift + y_shift
+
+
+def _sequential_logmatmulexp(logits):
+    """
+    For a tensor ``x`` whose time dimension is -3, computes::
+
+        x[..., 0, :, :] @ x[..., 1, :, :] @ ... @ x[..., T-1, :, :]
+
+    but does so numerically stably in log space.
+    """
+    batch_shape = logits.shape[:-3]
+    state_dim = logits.size(-1)
+    while logits.size(-3) > 1:
+        time = logits.size(-3)
+        even_time = time // 2 * 2
+        even_part = logits[..., :even_time, :, :]
+        x_y = even_part.reshape(batch_shape + (even_time // 2, 2, state_dim, state_dim))
+        x, y = x_y.unbind(-3)
+        contracted = _logmatmulexp(x, y)
+        if time > even_time:
+            contracted = torch.cat((contracted, logits[..., -1:, :, :]), dim=-3)
+        logits = contracted
+    return logits.squeeze(-3)
+
+
+class DiscreteHMM(TorchDistribution):
+    """
+    Hidden Markov Model with discrete latent state and arbitrary observation
+    distribution. This uses [1] to parallelize over time, achieving
+    O(log(time)) parallel complexity.
+
+    **References:**
+
+    [1] Simo Sarkka, Angel F. Garcia-Fernandez (2019)
+        "Temporal Parallelization of Bayesian Filters and Smoothers"
+        https://arxiv.org/pdf/1905.13002.pdf
+
+    :param torch.Tensor initial_logits: A logits tensor for an initial
+        categorical distribution over latent states. Should have rightmost size
+        ``state_dim`` and be broadcastable to ``batch_shape + (state_dim,)``.
+    :param torch.Tensor transition_logits: A logits tensor for transition
+        conditional distributions between latent states. Should have rightmost
+        shape ``(num_steps, state_dim, state_dim)`` (time, old, new), and be
+        broadcastable to ``batch_shape + (num_steps, state_dim, state_dim)``.
+    :param torch.distriburtions.Distribution observation_dist: A conditional
+        distribution of observed data conditioned on latent state. The
+        ``.batch_shape`` should have rightmost size ``state_dim`` and be
+        broadcastable to ``batch_shape + (num_steps + state_dim)``. The
+        ``.event_shape`` may be arbitrary.
+    """
+    arg_constraints = {"initial_logits": constraints.real,
+                       "transition_logits": constraints.real}
+
+    def __init__(self, initial_logits, transition_logits, observation_dist, validate_args=None):
```

Comment:
is `validate_args=None` more confusing than `validate_args=False`?

Reply:
`validate_args=None` has the special meaning "consult the global validation flag", which is controlled by `enable_validation()`. This behavior can be overridden in a single distribution by passing either `True` or `False`.